File: Algorithm/Sorting/Count_Sort/Count_Sort.cpp
#include <iostream>
#include <limits>
#include <vector>
using namespace std;

void count_sort(vector<int> &v)
{
    int n = v.size();

    int max_value = numeric_limits<int>::min();

    for (int i = 0; i < n; i++)
    {
        max_value = max(max_value, v[i]);
    }

    vector<int> frq(max_value + 1, 0);

    for (int i = 0; i < n; i++)
    {
        frq[v[i]]++;
    }

    for (int i = 1; i <= max_value; i++)
    {
        frq[i] += frq[i - 1];
    }

    vector<int> temp(n);

    for (int i = n - 1; i >= 0; i--)
    {
        // result[Position] = value
        // Pre-Decrement, it is important
        temp[--frq[v[i]]] = v[i];
    }

    for (int i = 0; i < n; i++)
    {
        v[i] = temp[i];
    }
}

int main()
{
    vector<int> v = {6, 2, 4, 1, 17, 5, 13};

    count_sort(v);

    for (int val : v)
        cout << val << " ";
    cout << endl;
}

File: Algorithm/Sorting/Count_Sort/Count_Sort.py
def count_sort(l, n):
    maxi = float("-inf")

    for i in range(n):
        maxi = max(l[i], maxi)

    count = [0] * (maxi + 1)

    for i in l:
        count[i] += 1

    for i in range(1, maxi + 1):
        count[i] += count[i - 1]

    output = [0] * n
    for i in range(n - 1, -1, -1):
        count[l[i]] -= 1
        output[count[l[i]]] = l[i]

    return output


def main():
    lst = [4, 4, 1, 0, 2, 2, 8, 9, 10]

    result = count_sort(lst, len(lst))

    for i in result:
        print(i)


if __name__ == "__main__":
    main()
